### ISSUE TITLE
31884 Bus error, how do I create tiles from the files calculating ~ 120 million features ? 

### ISSUE THREAD

**Issue by Amit-Gore** [mapbox/tippecanoe#648]
Hi,
I am trying to create the tiles from the geojson source(~250 gb) and I have enough of disk space(2T) for these tiles. But Tippecanoe is giving me 31884 Bus error along with the count of 120 million features. And it won't start creating the tiles.
How do I handle this much of data for tile creation? Any suggestions?

**Comment by Amit-Gore** [comment]
Yes, it works correctly with the same chunk of data. I am using following options while running
`--maximum-zoom=14 --detect-longitude-wraparound --no-tiny-polygon-reduction --no-tile-size-limit --no-line-simplification --no-tile-compression --include="A" --include="B" --include="C’" --include="D" --include="E" --include="F" --include="G"`
Can you assist me with running the tippecanoe under gdb ? How do I run it in background as the process takes nearly a day?

**Comment by ericfischer** [comment]
OK, the `terminate called after throwing an instance of 'std::bad_alloc' what(): std::bad_alloc` clarifies a lot: the Tippecanoe process is running out of memory.

This is probably because you have done `--no-tiny-polygon-reduction` and `--no-line-simplification`, so it has to try to represent every polygon, no matter how small, in all of its facets, in the zoom level 0 tile. I would suggest running it again without these options.

If you want to run it under gdb, you would do:

```
$ gdb tippecanoe
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
…
Reading symbols from tippecanoe...done.
(gdb) run --maximum-zoom=14 --detect-longitude-wraparound --no-tiny-polygon-reduction --no-tile-size-limit --no-line-simplification --no-tile-compression --include="A" --include="B" --include="C’" --include="D" --include="E" --include="F" --include="G"
Starting program: /usr/local/bin/tippecanoe --maximum-zoom=14 --detect-longitude-wraparound --no-tiny-polygon-reduction --no-tile-size-limit --no-line-simplification --no-tile-compression --include="A" --include="B" --include="C’" --include="D" --include="E" --include="F" --include="G"
```

using your regular command line arguments as options to the gdb `run` command.

I don't think there is a way to run a program under gdb in the background. I think it gets foregrounded as soon as you attach the debugger to the process. But you could run it in `screen` or  `tmux` so that the entire login session is backgrounded.

**Comment by Amit-Gore** [comment]
Ok so I tried removing those options over a chunk of data but it hampers the precision at higher zoom levels. I can set minimum-zoom to 6 to help to avoid the memory issue to a certain extent here. Is there any other way you can suggest to avoid this memory issue and make it run successfully?

**Comment by ericfischer** [comment]
You can use `--simplify-only-low-zooms` instead of `--no-line-simplification` to simplify only at the low zoom levels and preserve all the points at the highest zoom level.

**Comment by Amit-Gore** [comment]
@ericfischer Hey Eric, I am facing few problems here. I am using the following command attributes while generating the tiles 
`--minimum-zoom=10 --maximum-zoom=14 --detect-longitude-wraparound --no-tiny-polygon-reduction --no-tile-size-limit --simplify-only-low-zooms --drop-densest-as-needed --extend-zooms-if-still-dropping --no-tile-compression`


During the read phase, it reads ~100 million records and suddenly gives 
`You will probably run out of disk space.
28294160485 bytes used or committed, of 31435988992 originally available
geom: Write to temporary file failed: No space left on device
`
error. 

I think it uses the root disk(31gb) for writing it temporarily. Is there any other way to change it to some other disk? And what exactly the library does in this "read" phase?

![image](https://user-images.githubusercontent.com/7140231/48418960-b7c67e00-e77c-11e8-8e47-039f5c8d267d.png)

**Comment by ericfischer** [comment]
Thanks for answering about the temporary file location, @geeknik!

To answer the other question: there are basically three phases to Tippecanoe processing:

* Reading, during which GeoJSON is parsed, reprojected, and converted to an internal format in a temporary file
* Sorting, during which all the features in the temporary file are reordered into quadkey order, to facilitate separation into tiles, as well as density calculation and dot-dropping
* Tiling, when the internal data format is split into tiles and converted into vector tile format

"Reading" creates the temporary file, "sorting" reorders it, and "tiling" reads through it once for each zoom level and at the same time creates a new one for the following zoom level.
